winspool/win32.go

```diff
@@ -0,0 +1,1465 @@
+	if err != nil {
+		return nil, err
+	}
+
+	hdr := reflect.SliceHeader{
+		Data: uintptr(unsafe.Pointer(&pPrinterEnum[0])),
+		Len:  int(pcReturned),
+		Cap:  int(pcReturned),
+	}
+	printers := *(*[]PrinterInfo4)(unsafe.Pointer(&hdr))
+
+	printerNames := make([]string, 0, len(printers))
+	for _, printer := range printers {
+		printerNames = append(printerNames, utf16PtrToString(printer.pPrinterName))
+	}
+
+	return printerNames, nil
+}
+
+func EnumPrinters5() (map[string]string, error) {
+	pPrinterEnum, pcReturned, err := enumPrinters(5)
+	if err != nil {
+		return nil, err
+	}
+
+	hdr := reflect.SliceHeader{
+		Data: uintptr(unsafe.Pointer(&pPrinterEnum[0])),
+		Len:  int(pcReturned),
+		Cap:  int(pcReturned),
+	}
+	printers := *(*[]PrinterInfo5)(unsafe.Pointer(&hdr))
+
+	m := make(map[string]string, len(printers))
+	for _, printer := range printers {
+		printerName := utf16PtrToString(printer.pPrinterName)
+		portName := utf16PtrToString(printer.pPortName)
+		m[printerName] = portName
+	}
+
+	return m, nil
+}
+
+type HANDLE uintptr
+
+func OpenPrinter(printerName string) (HANDLE, error) {
+	var pPrinterName *uint16
+	pPrinterName, err := syscall.UTF16PtrFromString(printerName)
+	if err != nil {
+		return 0, err
+	}
+
+	var hPrinter HANDLE
+	_, _, err = openPrinterProc.Call(uintptr(unsafe.Pointer(pPrinterName)), uintptr(unsafe.Pointer(&hPrinter)), 0)
+	if err != NO_ERROR {
+		return 0, err
+	}
+	return hPrinter, nil
+}
+
+func (hPrinter *HANDLE) ClosePrinter() error {
+	_, _, err := closePrinterProc.Call(uintptr(*hPrinter))
+	if err != NO_ERROR {
+		return err
+	}
+	*hPrinter = 0
+	return nil
+}
+
+func (hPrinter HANDLE) getPrinter(level uint32) ([]byte, error) {
+	var cbBuf uint32
+	_, _, err := getPrinterProc.Call(uintptr(hPrinter), uintptr(level), 0, 0, uintptr(unsafe.Pointer(&cbBuf)))
+	if err != ERROR_INSUFFICIENT_BUFFER {
+		return nil, err
+	}
+
+	var pPrinter []byte = make([]byte, cbBuf)
+	_, _, err = getPrinterProc.Call(uintptr(hPrinter), uintptr(level), uintptr(unsafe.Pointer(&pPrinter[0])), uintptr(cbBuf), uintptr(unsafe.Pointer(&cbBuf)))
+	if err != NO_ERROR {
+		return nil, err
+	}
+
+	return pPrinter, nil
+}
+
+func (hPrinter HANDLE) GetPrinter5() (string, string, error) {
+	pPrinter, err := hPrinter.getPrinter(5)
+	if err != nil {
+		return "", "", err
+	}
+	var printerInfo5 PrinterInfo5 = *(*PrinterInfo5)(unsafe.Pointer(&pPrinter[0]))
+
+	printerName := utf16PtrToString(printerInfo5.pPrinterName)
+	portName := utf16PtrToString(printerInfo5.pPortName)
+
+	return printerName, portName, nil
+}
+
+func (hPrinter HANDLE) GetPrinter8() (*DevMode, error) {
+	pPrinter, err := hPrinter.getPrinter(8)
+	if err != nil {
+		return nil, err
+	}
+	var printerInfo8 PrinterInfo89 = *(*PrinterInfo89)(unsafe.Pointer(&pPrinter[0]))
+	var devMode *DevMode = (*DevMode)(unsafe.Pointer(printerInfo8.pDevMode))
+
+	return devMode, nil
+}
+
+func (hPrinter HANDLE) EnumPrinterDataEx() (map[string]interface{}, error) {
```

Comment:
There are several such functions in this file. They are Windows API wrappers, so I chose to leave them for future use.